enhancements/ingress/gateway-api-with-cluster-ingress-operator.md

```diff
@@ -0,0 +1,434 @@
+"openshift-ingress-operator" namespace, alongside its IngressController.
+
+To add DNS management for Gateway API, the DNS controller must allow a DNSRecord
+to be associated with a Gateway (not an IngressController) and be in the same
+namespace as its Gateway.  This is a relatively minor change.
+
+#### Controller to manage Gateway API CRDs and OpenShift Service Mesh
+
+This enhancement adds a new controller to do the following:
+
+1. Create the gatewayclasses CRD.
+2. Reconcile GatewayClasses by installing OSSM and other CRDs when appropriate.
+3. Reconcile Gateways by configuring OSSM, services, and DNS.
+
+This controller creates the gatewayclasses CRD when it starts, which enables the
+cluster admin to create a new GatewayClass.  When the cluster admin does this
+and specifies a controller name that OpenShift owns, the Ingress Operator then
+does the following:
+
+1. Create a Subscription to install OSSM.
+2. Create a ServiceMeshControlPlane to configure OSSM.
+3. Create a LoadBalancer-type service (or OSSM may create it; see [Open Questions](#open-questions)).
+4. Create a DNSRecord for the service.
+
+When OSSM is installed, it deploys Istiod and Envoy.  The Ingress Operator
+configures the ServiceMeshControlPlane to enable Gateway API support, which
+means that Istiod watches Gateway and HTTPRoute objects and configures Envoy
+accordingly.  Configuring a load balancer and DNS enables clients to connect
+through Envoy to backend applications as specified by the Gateway and HTTPRoute
+objects.
+
+### Workflow Description
+
+1. The cluster admin creates a GatewayClass with `spec.controllerName: openshift.io/gateway-controller`.
+2. The cluster admin creates a TLS secret in the "openshift-ingress" namespace with a `*.gwapi.example.com` wildcard default certificate.
+3. The cluster admin creates a Gateway, specifying the desired listeners, the secret from Step 2, and the GatewayClass from Step 1.
+4. The project admin creates an HTTPRoute, specifying the desired host name, the Gateway from Step 3, and the target service in the project admin's namespace.
+5. The service end-user connects to the host name specified in the HTTPRoute from Step 4 and receives a response from the service that the HTTPRoute targets.
+
+### API Extensions
+
+This enhancement imports the v1beta1 versions of the following CRDs into OpenShift:
+
+* [GatewayClass](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/).
+* [Gateway](https://gateway-api.sigs.k8s.io/api-types/gateway/).
+* [HTTPRoute](https://gateway-api.sigs.k8s.io/api-types/httproute/).
+
+Before this enhancement, DNSRecord CRs were not reconciled unless they were in
+the "openshift-ingress-operator" namespace and had an associated
+IngressController.  This enhancement enables reconciliation of DNSRecord CRs in
+the "openshift-ingress" namespace.
+
+This enhancement adds a controller that creates DNSRecord CRs for specially
+annotated services in the "openshift-ingress" namespace.  This controller is
```

Comment:
How does this annotation get on the service? Do cluster-admins have to manually annotate the service? Or is the C-I-O somehow annotating the service from derived information? Could we talk to that a bit here?

Have we considered External DNS's design https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/gateway-api.md in which they utilize the hostname on the `xRoute` objects. And for UDPRoute and TCPRoute that don't use hostnames in the spec, you add an annotation?

Reply:
> How does this annotation get on the service? 

C-I-O adds it.  I'll elaborate in the text.  

> Have we considered External DNS's design https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/gateway-api.md in which they utilize the hostname on the `xRoute` objects. And for UDPRoute and TCPRoute that don't use hostnames in the spec, you add an annotation?

Yes; this is [NE-1016](https://issues.redhat.com/browse/NE-1016).  I'd like for us to explore this option, but we don't have capacity to prioritize it.

Reply:
I'll mention NE-1016 in "Alternatives" though.

Reply:
Actually, using the annotation doesn't make sense if we want a DNS record for each Gateway listener host name, and why wouldn't we want that?  It makes more sense to have cluster-ingress-operator watch gateways and create a DNSRecord CR for each host name specified in any listener for which the Gateway specifies the istio-ingressgateway service.  This complicates the controller logic, but not too much, I think.  What do you think?